File: sdks/python/roost/resources/installer.py
"""``roost.installer`` — agent installer binary management (superadmin).

Wraps ``/api/installer*``. ``upload()`` is the canonical 3-step flow: request a
signed PUT url, stream the binary to it, finalize. One ``Idempotency-Key``
spans request+finalize so retries replay while the signed url is still valid.
"""

from __future__ import annotations

import hashlib
import os
import uuid
from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING, Any

import httpx

if TYPE_CHECKING:
    from roost.client import RoostClient


@dataclass(slots=True)
class InstallerFile:
    """One platform's installer within a version."""

    download_url: str
    checksum_sha256: str | None
    file_size: int | None
    file_name: str | None
    uploaded_at: int | None


@dataclass(slots=True)
class InstallerVersion:
    version: str
    # the windows installer; an alias of files["windows_x64"].download_url
    download_url: str | None
    checksum_sha256: str | None
    release_notes: str | None
    file_size: int | None
    uploaded_at: int | None
    uploaded_by: str | None
    deleted_at: int | None
    release_date: str | None = None
    promoted_at: int | None = None
    promoted_by: str | None = None
    # keyed windows_x64 / macos_arm64 / linux_x64; a version uploaded before
    # per-platform files carries its windows installer under windows_x64
    files: dict[str, InstallerFile] = field(default_factory=dict)


def _parse_file(raw: dict[str, Any]) -> InstallerFile:
    return InstallerFile(
        download_url=str(raw.get("download_url", "")),
        checksum_sha256=raw.get("checksum_sha256"),
        file_size=raw.get("file_size"),
        file_name=raw.get("file_name"),
        uploaded_at=raw.get("uploaded_at"),
    )


def _parse_version(raw: dict[str, Any]) -> InstallerVersion:
    files = raw.get("files")
    return InstallerVersion(
        version=str(raw.get("version", "")),
        download_url=raw.get("download_url"),
        checksum_sha256=raw.get("checksum_sha256"),
        release_notes=raw.get("release_notes"),
        file_size=raw.get("file_size"),
        uploaded_at=raw.get("uploaded_at"),
        uploaded_by=raw.get("uploaded_by"),
        deleted_at=raw.get("deletedAt"),
        release_date=raw.get("release_date"),
        promoted_at=raw.get("promoted_at"),
        promoted_by=raw.get("promoted_by"),
        files={
            str(k): _parse_file(v)
            for k, v in (files.items() if isinstance(files, dict) else ())
            if isinstance(v, dict)
        },
    )


class Installer:
    """Installer binary metadata + upload flow (superadmin-only)."""

    def __init__(self, client: "RoostClient") -> None:
        self._client = client
        # Test hook: set an httpx.AsyncBaseTransport to intercept the signed-url PUT.
        self._upload_transport: httpx.AsyncBaseTransport | None = None

    async def list(
        self,
        *,
        include_deleted: bool = False,
        page_size: int | None = None,
        page_token: str | None = None,
    ) -> dict[str, Any]:
        """List uploaded installer versions, newest first."""
        query: dict[str, Any] = {}
        if include_deleted:
            query["includeDeleted"] = True
        if page_size is not None:
            query["page_size"] = page_size
        if page_token:
            query["page_token"] = page_token
        resp = await self._client.request(
            "/api/installer",
            query=query or None,
        )
        data = resp.data if isinstance(resp.data, dict) else {}
        versions = [
            _parse_version(v)
            for v in (data.get("versions") or [])
            if isinstance(v, dict)
        ]
        return {
            "versions": versions,
            "next_page_token": str(data.get("next_page_token") or data.get("nextPageToken") or ""),
        }

    async def latest(self) -> InstallerVersion:
        """Return the current latest installer version metadata."""
        resp = await self._client.request("/api/installer/latest")
        data = resp.data if isinstance(resp.data, dict) else {}
        return _parse_version(data)

    async def upload(
        self,
        file_path: str | os.PathLike[str],
        *,
        version: str,
        platform: str | None = None,
        release_notes: str | None = None,
        set_as_latest: bool = True,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        """Upload an installer binary: POST for a signed url → PUT bytes → PUT finalize.

        ``platform`` is ``windows_x64``, ``macos_arm64`` or ``linux_x64``; the api
        derives it from the file extension (``.exe`` / ``.pkg`` / ``.deb``) when
        omitted and rejects a mismatch. One ``Idempotency-Key`` covers the POST
        and the finalize PUT.
        """
        path = Path(file_path)
        binary = path.read_bytes()
        file_size = len(binary)
        checksum = hashlib.sha256(binary).hexdigest()
        file_name = path.name

        idem = idempotency_key or f"py-sdk-installer-upload-{uuid.uuid4()}"

        # step 1: request signed upload url
        start_body: dict[str, Any] = {
            "version": version,
            "fileName": file_name,
            "setAsLatest": set_as_latest,
        }
        if platform is not None:
            start_body["platform"] = platform
        if release_notes is not None:
            start_body["releaseNotes"] = release_notes

        start_resp = await self._client.request(
            "/api/installer/upload",
            method="POST",
            body=start_body,
            idempotency_key=idem,
        )
        start_data = start_resp.data if isinstance(start_resp.data, dict) else {}
        upload_url = start_data.get("uploadUrl")
        upload_id = start_data.get("uploadId")
        if not upload_url or not upload_id:
            msg = "installer.upload: server response missing uploadUrl or uploadId"
            raise RuntimeError(msg)

        # step 2: PUT bytes. One-shot client so the SDK's Authorization /
        # Roost-Version headers don't leak — the signed url rejects extras.
        client_kwargs: dict[str, Any] = {}
        if self._upload_transport is not None:
            client_kwargs["transport"] = self._upload_transport
        async with httpx.AsyncClient(**client_kwargs) as raw:
            put_resp = await raw.put(
                str(upload_url),
                content=binary,
                headers={
                    "Content-Type": "application/octet-stream",
                    "Content-Length": str(file_size),
                },
            )
        if put_resp.status_code >= 400:
            msg = (
                f"installer.upload: signed PUT failed "
                f"(status={put_resp.status_code}, body={put_resp.text[:200]!r})"
            )
            raise RuntimeError(msg)

        # step 3: finalize
        finalize_resp = await self._client.request(
            "/api/installer/upload",
            method="PUT",
            body={"uploadId": upload_id, "checksum_sha256": checksum},
            idempotency_key=idem,
        )
        return finalize_resp.data if isinstance(finalize_resp.data, dict) else {}

    async def set_latest(
        self,
        version: str,
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        resp = await self._client.request(
            f"/api/installer/{version}/set-latest",
            method="POST",
            body={},
            idempotency_key=idempotency_key,
        )
        return resp.data if isinstance(resp.data, dict) else {}

    async def delete(
        self,
        version: str,
        *,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        # Resource-specific prefix, not the core client's generic py-sdk key.
        idem = idempotency_key or f"py-sdk-installer-delete-{uuid.uuid4()}"
        resp = await self._client.request(
            f"/api/installer/{version}",
            method="DELETE",
            headers={"Idempotency-Key": idem},
        )
        return resp.data if isinstance(resp.data, dict) else {}


__all__ = ["Installer", "InstallerFile", "InstallerVersion"]

File: sdks/node/src/resources/installer.ts
/**
 * `owlette.installer` — superadmin-only installer-version management over the
 * wave-1B installer-api routes: GET /api/installer, GET /api/installer/latest,
 * POST+PUT /api/installer/upload, POST /api/installer/{version}/set-latest,
 * DELETE /api/installer/{version}.
 *
 * `upload` is a 3-step flow: POST for a signed URL → PUT the binary → PUT to
 * finalize. The sha-256 is computed client-side and handed to finalize so the
 * server can reject a corrupted upload.
 *
 * Mutations auto-generate `Idempotency-Key` when the caller omits one. Errors
 * surface as `OwletteApiError`.
 */
import { createHash, randomUUID } from 'crypto';
import { createReadStream } from 'fs';
import { stat } from 'fs/promises';
import type { OwletteClient } from '../lib/client';

export type InstallerPlatform = 'windows_x64' | 'macos_arm64' | 'linux_x64';

/** One platform's installer within a version. */
export interface InstallerFile {
  download_url: string;
  checksum_sha256: string | null;
  file_size: number | null;
  /** Display name, e.g. `Owlette-Installer-v3.4.0.pkg`. */
  file_name: string | null;
  uploaded_at: number | null;
}

export interface InstallerVersion {
  version: string;
  /** The windows installer; an alias of `files.windows_x64.download_url`. */
  download_url: string | null;
  checksum_sha256: string | null;
  release_notes: string | null;
  file_size: number | null;
  uploaded_at: number | null;
  uploaded_by: string | null;
  release_date?: string | null;
  deletedAt: number | null;
  /** The installer per platform. A version uploaded before per-platform
   * files carries its windows installer under `windows_x64`. */
  files: Partial<Record<InstallerPlatform, InstallerFile>>;
  promoted_at?: number | null;
  promoted_by?: string | null;
}

export interface ListInstallerOptions {
  pageSize?: number;
  pageToken?: string;
  includeDeleted?: boolean;
}

export interface ListInstallerResult {
  versions: InstallerVersion[];
  nextPageToken: string;
}

export interface UploadRequestOptions {
  /** Local path to the `.exe`, `.pkg` or `.deb` to upload. */
  filePath: string;
  /** Semver `X.Y.Z`. */
  version: string;
  /** File name on storage; defaults to the basename of `filePath`. */
  fileName?: string;
  /** Platform key the file is stored under; the api derives it from the file
   * name's extension when omitted and rejects a mismatch. */
  platform?: InstallerPlatform;
  contentType?: string;
  releaseNotes?: string | null;
  /** Whether to promote this version to `latest` on finalize. Default true. */
  setAsLatest?: boolean;
  /** Override the auto-generated idempotency key used on both API calls. */
  idempotencyKey?: string;
  /** @deprecated Use idempotencyKey; retained for compatibility. */
  startIdempotencyKey?: string;
  /** @deprecated Use idempotencyKey; retained for compatibility. */
  finalizeIdempotencyKey?: string;
  /** Override the upload PUT; defaults to `client._fetch` so a custom proxy/agent
   * passed at construction time is honoured. */
  signedUploadFetch?: typeof fetch;
}

export interface UploadResult {
  version: string;
  download_url: string;
  checksum_sha256: string;
  file_size: number;
  platform: InstallerPlatform;
  /** Every platform file the version carries after this upload. */
  files: Partial<Record<InstallerPlatform, InstallerFile>>;
}

export class Installer {
  constructor(private readonly client: OwletteClient) {}

  async list(opts: ListInstallerOptions = {}): Promise<ListInstallerResult> {
    const query: Record<string, string | number | boolean | undefined> = {};
    if (opts.pageSize !== undefined) query.page_size = opts.pageSize;
    if (opts.pageToken !== undefined) query.page_token = opts.pageToken;
    if (opts.includeDeleted) query.includeDeleted = 'true';

    const res = await this.client.request<{
      versions: InstallerVersion[];
      next_page_token?: string;
      nextPageToken?: string;
    }>('/api/installer', { query });
    return {
      versions: res.data.versions ?? [],
      nextPageToken: res.data.next_page_token ?? res.data.nextPageToken ?? '',
    };
  }

  async latest(): Promise<InstallerVersion> {
    const res = await this.client.request<InstallerVersion>('/api/installer/latest');
    return res.data;
  }

  /**
   * Three-step upload: request a signed URL, PUT the binary, finalize. Streams
   * the file once for its sha-256. Returns the finalized version metadata.
   */
  async upload(opts: UploadRequestOptions): Promise<UploadResult> {
    const fileName =
      opts.fileName ?? opts.filePath.split(/[\\/]/).pop() ?? 'installer.exe';
    const contentType = opts.contentType ?? 'application/octet-stream';

    // 1. POST /api/installer/upload — request the signed URL.
    const startBody: Record<string, unknown> = {
      version: opts.version,
      fileName,
      contentType,
    };
    if (opts.platform !== undefined) startBody.platform = opts.platform;
    if (opts.releaseNotes !== undefined) startBody.releaseNotes = opts.releaseNotes;
    if (opts.setAsLatest !== undefined) startBody.setAsLatest = opts.setAsLatest;
    const idempotencyKey =
      opts.idempotencyKey ??
      opts.startIdempotencyKey ??
      opts.finalizeIdempotencyKey ??
      `sdk-installer-upload-${randomUUID()}`;

    const startRes = await this.client.request<{
      uploadUrl: string;
      uploadId: string;
      storagePath: string;
      expiresAt: string;
    }>('/api/installer/upload', {
      method: 'POST',
      body: startBody,
      idempotencyKey: opts.startIdempotencyKey ?? idempotencyKey,
    });
    const { uploadUrl, uploadId } = startRes.data;

    // 2. PUT the signed URL — upload the binary.
    const fileStat = await stat(opts.filePath);
    const fileSize = fileStat.size;
    const sha256 = await sha256OfFile(opts.filePath);

    const buffer = await readFileToBuffer(opts.filePath);
    const uploadFetch = opts.signedUploadFetch ?? this.client._fetch;
    const putRes = await uploadFetch(uploadUrl, {
      method: 'PUT',
      headers: { 'Content-Type': contentType },
      body: new Uint8Array(buffer),
    });
    if (!putRes.ok) {
      throw new Error(
        `installer.upload: signed-url PUT failed with status ${putRes.status} ` +
          `(${fileSize} bytes from ${opts.filePath})`,
      );
    }

    // 3. PUT /api/installer/upload — finalize.
    const finalizeRes = await this.client.request<UploadResult>(
      '/api/installer/upload',
      {
        method: 'PUT',
        body: { uploadId, checksum_sha256: sha256 },
        idempotencyKey: opts.finalizeIdempotencyKey ?? idempotencyKey,
      },
    );
    return finalizeRes.data;
  }

  async setLatest(
    version: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<{ version: string; latest: Record<string, unknown> }> {
    const res = await this.client.request<{
      version: string;
      latest: Record<string, unknown>;
    }>(`/api/installer/${encodeURIComponent(version)}/set-latest`, {
      method: 'POST',
      body: {},
      idempotencyKey: opts.idempotencyKey ?? `sdk-installer-set-latest-${randomUUID()}`,
    });
    return res.data;
  }

  async delete(
    version: string,
    opts: { idempotencyKey?: string } = {},
  ): Promise<{ version: string; deletedAt: number; alreadyDeleted: boolean }> {
    const res = await this.client.request<{
      version: string;
      deletedAt: number;
      alreadyDeleted: boolean;
    }>(`/api/installer/${encodeURIComponent(version)}`, {
      method: 'DELETE',
      idempotencyKey: opts.idempotencyKey ?? `sdk-installer-delete-${randomUUID()}`,
    });
    return res.data;
  }
}

async function sha256OfFile(filePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const hash = createHash('sha256');
    const stream = createReadStream(filePath);
    stream.on('data', (chunk) => hash.update(chunk));
    stream.on('end', () => resolve(hash.digest('hex')));
    stream.on('error', reject);
  });
}

async function readFileToBuffer(filePath: string): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    const stream = createReadStream(filePath);
    stream.on('data', (chunk) => chunks.push(chunk as Buffer));
    stream.on('end', () => resolve(Buffer.concat(chunks)));
    stream.on('error', reject);
  });
}
